File: typescript/exercises.ts
import { PathLike } from "node:fs"
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  items: T[],
  predicate: (item: T) => boolean,
  consumer: (item: T) => U
): U | undefined {
  const item = items.find(predicate)
  return item ? consumer(item) : undefined
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power
  }
}

export async function meaningfulLineCount(filePath: PathLike): Promise<number> {
  let count: number = 0
  const file = await open(filePath, "r")
  for await (const line of file.readLines()) {
    const trimmed = line.trim()
    if (trimmed && !trimmed.startsWith("#")) {
      count++
    }
  }
  return count
}

interface Box {
  kind: "Box"
  readonly width: number
  readonly depth: number
  readonly length: number
}

interface Sphere {
  kind: "Sphere"
  readonly radius: number
}

export type Shape = Box | Sphere

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box": {
      return shape.length * shape.width * shape.depth
    }
    case "Sphere": {
      return (4 / 3) * Math.PI * shape.radius ** 3
    }
  }
}

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box": {
      return (
        2 *
        (shape.length * shape.depth +
          shape.length * shape.width +
          shape.width * shape.depth)
      )
    }
    case "Sphere": {
      return 4 * Math.PI * shape.radius ** 2
    }
  }
}

export interface BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  size(): number
  inorder(): Generator<T>
  toString(): string
}

export class Empty<T> implements BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty<T>(), new Empty<T>())
  }

  contains(value: T): boolean {
    return false
  }

  size(): number {
    return 0
  }

  *inorder(): Generator<T> {
    return
  }

  toString(): string {
    return "()"
  }
}

class Node<T> implements BinarySearchTree<T> {
  constructor(
    private value: T,
    private left: BinarySearchTree<T> = new Empty<T>(),
    private right: BinarySearchTree<T> = new Empty<T>()
  ) {}

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right)
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value))
    } else {
      return this
    }
  }

  contains(value: T): boolean {
    if (value < this.value) {
      return this.left.contains(value)
    } else if (value > this.value) {
      return this.right.contains(value)
    } else {
      return true
    }
  }

  size(): number {
    return 1 + this.left.size() + this.right.size()
  }

  *inorder(): Generator<T> {
    yield* this.left.inorder()
    yield this.value
    yield* this.right.inorder()
  }

  toString(): string {
    const leftStr = this.left.toString() === "()" ? "" : this.left.toString()
    const rightStr = this.right.toString() === "()" ? "" : this.right.toString()
    return `(${leftStr}${this.value}${rightStr})`
  }
}

File: haskell/Exercises.hs
module Exercises
    ( change,
      firstThenApply,
      powers,
      meaningfulLineCount,
      Shape(..),
      volume,
      surfaceArea,
      BST(Empty),
      size,
      inorder,
      insert,
      contains,
      show
    ) where

import qualified Data.Map as Map
import Data.Text (pack, unpack, replace)
import Data.List(isPrefixOf, find)
import Data.Char(isSpace)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
    | amount < 0 = Left "amount cannot be negative"
    | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
        where
          changeHelper [] remaining counts = counts
          changeHelper (d:ds) remaining counts =
            changeHelper ds newRemaining newCounts
              where
                (count, newRemaining) = remaining `divMod` d
                newCounts = Map.insert d count counts

firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs pred f = f <$> find pred xs

powers :: Integral a => a -> [a]
powers base = map (base^) [0..]

meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount path = do
  contents <- readFile path
  return $ length $ filter meaningfulLine $ lines contents
  where
    meaningfulLine line =
      not (all isSpace line) && not ("#" `isPrefixOf` dropWhile isSpace line)

data Shape
    = Sphere Double 
    | Box Double Double Double 
    deriving (Eq, Show) 

volume :: Shape -> Double
volume (Sphere r) = (4/3) * pi * r^3
volume (Box l w h) = l * w * h

surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4 * pi * r^2
surfaceArea (Box l w h) = 2 * (l * w + l * h + w * h)

data BST a
    = Empty
    | Node a (BST a) (BST a)

size :: BST a -> Int
size Empty = 0
size (Node _ left right) = 1 + size left + size right

inorder :: BST a -> [a]
inorder Empty = []
inorder (Node value left right) = inorder left ++ [value] ++ inorder right

insert :: Ord a => a -> BST a -> BST a
insert value Empty = Node value Empty Empty
insert value (Node nodeValue left right)
    | value < nodeValue = Node nodeValue (insert value left) right
    | value > nodeValue = Node nodeValue left (insert value right)
    | otherwise = Node nodeValue left right

contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains target (Node value left right) 
    | target == value = True
    | target < value  = contains target left
    | target > value  = contains target right

instance Show a => Show (BST a) where
    show :: Show a => BST a -> String
    show Empty = "()"
    show (Node value left right) = 
        "(" ++ leftStr ++ show value ++ rightStr ++ ")"
      where
        leftStr = case left of
            Empty -> ""
            _     -> show left
        rightStr = case right of
            Empty -> ""
            _     -> show right